Tuples typechecking, evaluation and render

EmiliaPaz opened the pull request:
Tuples are implemented as:

```
Term = ... | Tuple Term Term
VType = ... | TTuple VType VType
Val = ... | VTuple Val Val 
```

It will typecheck as long as it is valid (if it holds a var that doesn't exist in the environment, it will not typecheck). Same with the evaluation.

I merged master into this branch to fix all the changes, so there is no conflict

Tested it with dummy input. Still need to parse it using Chevrotain.

Commit:
Tuples working with type (Tuple Term Term)

Commit:
Merge remote-tracking branch 'origin' into tuples

Commit:
Merge master to tuples, without new type

Commit:
Tuples typecheck, evaluation and render. Need to be parsed

alexshadley reviewed (changes requested):
First of all, good work on the PR!  All of the code you've written looks great; good intuition that we need a term, type and value for tuples.  I have a few comments on what we might add.

First, we might also want a `fst` and `snd` operator to get the first and second element of a tuple -- otherwise after we build a tuple there's no way to access the elements within.  And second, having some tests for tuples would be great.

Additionally, would you feel comfortable writing the parsing code for tuples?  I think this may be a good problem for learning how Chevrotain works.  If you'd like I could help walk you through the process on Wednesday, I understand that Chevrotain can be tricky to work with, and on top of that you have to contend with some arbitrary choices I made along my relatively inexperienced implementation.

brian-quiroz commented on src/Types.elm:
```diff
@@ -71,6 +71,7 @@ listToTypeSign types =
+    TBool           -> [TBool]
+    TInt            -> [TInt]
+    TFun vt1 vt2    -> (typeSignToList vt1) ++ (typeSignToList vt2)
+    TTuple vt1 vt2  -> (typeSignToList vt1) ++ (typeSignToList vt2)   -- Will return the tuple as a list. Not ideal, but the function is type sign to list
```
I think `typeSignToList (TTuple vt1 vt2 )` should return `[TTuple vt1 vt2 ]`.

The point of the typeSignToList function was to get the individual types into list form so that they're easier to work with. So if we have `TFun (TInt) (TFun TInt TBool)` this would be converted to `[TInt, TInt, TBool]` so that we can easily  work with the components of the type signature (for example, having this vtype in list form makes it easier print this function type as `TInt -> TInt -> TBool`).

Having said that I think it would make more sense to have this case return a list containing `[VTuple vt1 vt2]`. Ultimately we'll have to refactor the way we're obtaining the type signature so I don't think it's a huge deal but for now we should probably change it.

brian-quiroz commented on src/Typecheck.elm:
```diff
@@ -151,6 +152,12 @@ typecheck env t =
+      Tuple a b ->
+        case (typecheck env a, typecheck env b) of
+          (Checks x, Checks y) -> Checks (TTuple x y)
+          _ -> Invalid
```
Even if `typecheck env a` or `typecheck env b` return `Partial _` or `Fails _ _ _ _` we will return `Invalid`. Is this the intended behavior? Wouldn't we want to cascade a `Partial` or `Fails` in one or both components of a tuple to the result of type checking that tuple?

EmiliaPaz replied:
You are right, it should cascade Partial or Fails. I'll work on this

EmiliaPaz replied:
I went ahead and changed the return types according to the typecheck of each member of the tuple. There are 12 different cases, and I was a little unsure of how to deal with them. At the end I implemented it according to the following table, which we can discuss and easily change if need to be:

Followed this convention (we can discuss this)
(x,y)

x\y         | Checks  | Fails    | Partial   | Invalid
- - - - - - - - - - - - - - - - - - - - - - - - - - - 
Checks | Checks | Fails     | Partial   | Invalid
Fails      | Fails      | Fails     | Fails      | Invalid
Partial   | Partial   | Fails     | Partial   | Invalid
Invalid   | Invalid   | Invalid  | Invalid   | Invalid

EmiliaPaz replied:
I will wait to merge this to master until we have discussed if the vtypes are appropriate

brian-quiroz reviewed (changes requested):
I think your code looks good! I just added some comments/concerns I had .

Commit:
TTuple typeSignToList change to return a list holding the TTuple

Commit:
Typecheck for tuples returns Checks, Fails, Partial or Invalid

Followed this convention (we can discuss this)
x\y         | Checks  | Fails    | Partial   | Invalid
-----------------------------------------------------
Checks | Checks | Fails     | Partial   | Invalid
Fails      | Fails      | Fails     | Fails      | Invalid
Partial   | Partial   | Fails     | Partial   | Invalid
Invalid   | Invalid   | Invalid  | Invalid   | Invalid

EmiliaPaz commented:
> First of all, good work on the PR! All of the code you've written looks great; good intuition that we need a term, type and value for tuples. I have a few comments on what we might add.
> 
> First, we might also want a `fst` and `snd` operator to get the first and second element of a tuple -- otherwise after we build a tuple there's no way to access the elements within. And second, having some tests for tuples would be great.
> 
> Additionally, would you feel comfortable writing the parsing code for tuples? I think this may be a good problem for learning how Chevrotain works. If you'd like I could help walk you through the process on Wednesday, I understand that Chevrotain can be tricky to work with, and on top of that you have to contend with some arbitrary choices I made along my relatively inexperienced implementation.

Commit:
Forgot to change environment variable for Tuple typechecking
